### ISSUE TITLE
Annotations vs. Logs

### ISSUE THREAD

**Issue by SergeyKanzhelev** [bogdandrutu/openconsensus#16]
Annotations and logs are equivalents. However, annotations are more structured, with an explicit description field.
OT Log
public class Fields {
    public static final String ERROR_KIND = "error.kind";
    public static final String ERROR_OBJECT = "error.object";
   // etc
}
Span.log(Map<String, ?> fields);
Span.log(long timestampMicroseconds, Map<String, ?> fields);
Span.log(long timestampMicroseconds, String event);

OC Annotation
public abstract class Annotation {
  public static Annotation fromDescription(String description);
  public static Annotation fromDescriptionAndAttributes(
      String description, Map<String, AttributeValue> attributes);
  public abstract String getDescription();
  public abstract Map<String, AttributeValue> getAttributes();
}

TODO: what to do with the extra description field? How is it exported?

**Comment by SergeyKanzhelev** [comment]
[ted] Idea of required string is great
[Yuri] Description may be a bad name. Perhaps Event with the string.

**Comment by SergeyKanzhelev** [comment]
[Yuri] EventName instead of description
[sergey] everybody agree `log` is confusing.
[ted] Annotation -> `Event` + `description` -> `name`

**Comment by SergeyKanzhelev** [comment]
Yay! closed

**Comment by yurishkuro** [comment]
Rationale: sometimes events are more semantically meaningful than just "log strings" or "description". For example, when tracing http request with Go (e.g. [this blog][1]), there is a well defined set of HTTP events attached to the span (the [ClientTrace interface][2]), e.g.
  * DNS start
  * DNS done
  * GotFirstResponseByte
  * ConnectStart, etc.

OpenTracing spec recommends an attribute `event` for all logs, which is essentially saying "attach a meaningful string to each event". In case of a plain info log added to the span that string can be a free form description, but in the above example it could be a collection of well-defined strings that may be used for some additional post-collection processing.

In the future it would be nice if we could support causality references (happens-before relationship) between events, e.g.

```
ev1 := tracer.event("did-x")
ev2 := tracer.event("did-y").happensAfter(ev1)
```

This may be necessary to reason about critical path execution because relying on the timestamps alone does not provide sufficiently strong causality under multi-threading conditions. But this is something for another day.

[1]: https://medium.com/opentracing/tracing-http-request-latency-in-go-with-opentracing-7cc1282a100a
[2]: https://golang.org/pkg/net/http/httptrace/#ClientTrace

**Comment by SergeyKanzhelev** [comment]
Curiously, methods which checks whether annotations needs to be recorded were already called 
`isRecordingEvents` and `setRecordEvents` =)

**Comment by SergeyKanzhelev** [comment]
closed via #26
